what4/src/What4/Expr/BoolMap.hs

```diff
@@ -0,0 +1,170 @@
+{-|
+Module      : What4.Expr.BoolMap
+Copyright   : (c) Galois Inc, 2019
+License     : BSD3
+Maintainer  : rdockins@galois.com
+
+Declares a datatype for representing n-way conjunctions or disjunctions
+in a way that efficently capures important algebraic
+laws like commutativity, associativity and resolution.
+-}
+
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE ViewPatterns #-}
+module What4.Expr.BoolMap
+  ( BoolMap
+  , var
+  , addVar
+  , fromVars
+  , combine
+  , Polarity(..)
+  , negatePolarity
+  , contains
+  , isInconsistent
+  , isNull
+  , BoolMapView(..)
+  , viewBoolMap
+  , traverseVars
+  , reversePolarities
+  , removeVar
+  ) where
+
+import           Control.Lens (_1, over)
+import           Data.Bits
+import           Data.Hashable
+import           Data.List (foldl')
+import           Data.List.NonEmpty (NonEmpty(..))
+import           Data.Kind (Type)
+import           Data.Map (Map)
+import qualified Data.Map.Strict as Map
+import qualified Data.Map.Merge.Strict as Map
+import           Data.Parameterized.Classes
+
+import           What4.BaseTypes
+
+-- | Describes the occurence of a variable or expression, whether it is
+--   negated or not.
+data Polarity = Positive | Negative
+ deriving (Eq,Ord,Show)
+
+instance Hashable Polarity where
+  hashWithSalt s Positive = hashWithSalt s (0::Int)
+  hashWithSalt s Negative = hashWithSalt s (1::Int)
+
+-- | Swap a polarity value
+negatePolarity :: Polarity -> Polarity
+negatePolarity Positive = Negative
+negatePolarity Negative = Positive
+
+newtype Wrap (f :: k -> Type) (x :: k) = Wrap { unWrap:: f x }
```

Comment:
This seems like something that might belong in parameterized-utils, in that it has little to do with SMT solvers, and is more generally about working with type-level stuff.

Reply:
Yeah, it's true... but I don't  have a good sense of what API it should have, etc. and the name really sucks.